pandas/core/reshape.py

```diff
@@ -940,24 +942,27 @@ def melt_stub(df, stub, i, j):
         varnames = get_var_names(df, "^" + stub)
         newdf = melt(df, id_vars=i, value_vars=varnames, value_name=stub,
                      var_name=j)
-        newdf_j = newdf[j].str.replace(stub, "")
-        try:
-            newdf_j = newdf_j.astype(int)
-        except ValueError:
-            pass
-        newdf[j] = newdf_j
-        return newdf
+        newdf[j] = Categorical(newdf[j])
+        newdf[j] = newdf[j].str.replace(stub, "")
+        return newdf.set_index([i, j])
 
     id_vars = get_var_names(df, "^(?!%s)" % "|".join(stubnames))
     if i not in id_vars:
         id_vars += [i]
 
-    newdf = melt_stub(df, stubnames[0], id_vars, j)
+    mstubs = [melt_stub(df, stub, i, j) for stub in stubnames]
+    mstubs = mstubs[0].join(mstubs[1:], how='outer')
+    new = df[id_vars].set_index(i).join(mstubs)
+
+    # The index of the new dataframe is [i, j], if the j column is a time
+    # variable, try to convert this to integer.
```

Comment:
Fixed, but regarding a character that separates the stub name from the variable part:

```python
In [7]: df = pd.DataFrame({ 'A.2010': np.random.rand(N),
   ...:                     'A.2011': np.random.rand(N),
   ...:                     'B.2010': np.random.rand(N),
   ...:                     'B.2011': np.random.rand(N),
   ...:                      'X' : np.random.randint(N, size=N),
   ...:     })
   ...:
   ...: df
   ...:
Out[7]:
     A.2010    A.2011    B.2010    B.2011  X
0  0.873404  0.467946  0.569808  0.358077  1
1  0.780154  0.554582  0.668437  0.810530  1
2  0.884003  0.555784  0.246305  0.038423  2
```

```python
In [8]: df['id'] = df.index
   ...: pd.wide_to_long(df, ['A.', 'B.'], i='id', j='year')
   ...:
Out[8]:
         X        A.        B.
id year
0  2010  1  0.873404  0.569808
1  2010  1  0.780154  0.668437
2  2010  2  0.884003  0.246305
0  2011  1  0.467946  0.358077
1  2011  1  0.554582  0.810530
2  2011  2  0.555784  0.038423
```

A user might expect the new separating character (`.`) to be stripped, like `reshape` in `R` does.

I.e. we could simply set `value_name=stub[:-1]` in the call to `melt_stub`?